lib/src/relations/to_one.dart

```diff
@@ -0,0 +1,126 @@
+import '../box.dart';
+import '../modelinfo/entity_definition.dart';
+import '../store.dart';
+
+/// Manages a to-one relation, an unidirectional link from a "source" entity to
+/// a "target" entity. The target object is referenced by its ID, which is
+/// persisted in the source object.
+///
+/// You can:
+///   - set [target]=null or [targetId]=0 to remove the relation.
+///   - set [target] to an object to set the relation.
+///     Call [Box<SourceEntity>.put()] to persist the changes. If the target
+///     object is a new one (its ID is 0), it will be also saved automatically.
+///   - set [targetId] to an existing object's ID to set the relation.
+///     Call [Box<SourceEntity>.put()] to persist the changes.
+///
+/// ```
+/// class Order: Entity {
+///   final customer = ToOne<Customer>();
+///   ...
+/// }
+///
+/// // Example 1: create a relation
+/// final order = Order(...);
+/// final customer = Customer();
+/// order.customer.target = customer;
+///
+/// // Or you could create the target object in place:
+/// // order.customer.target = Customer()
+///
+/// // attach() must be called when creating new instances. On objects (e.g.
+/// // "orders" in this example) read with box.get() its done automatically.
+/// order.customer.attach(store);
+///
+/// // saves both [customer] and [order] in the database
+/// store.box<Order>().put(order);
+///
+///
+/// // Example 2: remove a relation
+///
+/// order.customer.target = null
+/// // ... or ...
+/// order.customer.targetId = 0
+/// ```
+class ToOne<EntityT> {
+  /*late final*/ Box<EntityT> _box;
+
+  /*late final*/
+  EntityDefinition<EntityT> _entity;
+
+  _ToOneValue<EntityT> _value = _ToOneValue<EntityT>.none();
+
+  EntityT /*?*/ get target {
+    if (_value._state == _ToOneState.lazy) {
+      final object = _box /*!*/ .get(_value._id);
+      _value = (object == null)
+          ? _ToOneValue<EntityT>.unresolvable(_value._id)
+          : _ToOneValue<EntityT>.stored(_value._id, object);
+    }
+    return _value._object;
+  }
+
+  set target(EntityT /*?*/ object) {
+    if (object == null) {
+      _value = _ToOneValue<EntityT>.none();
+    } else {
+      final id = _entity.getId(object) ?? 0;
+      _value = (id == 0)
+          ? _ToOneValue<EntityT>.unstored(object)
+          : _ToOneValue<EntityT>.stored(id, object);
+    }
+  }
+
+  int get targetId => _value._id;
+
+  set targetId(int /*?*/ id) {
+    id ??= 0;
+    if (id == _value._id) return;
+    if (id == 0) {
+      _value = _ToOneValue<EntityT>.none();
+    } else if (_value._state == _ToOneState.unstored &&
+        id == _entity.getId(_value._object)) {
```

Comment:
Can this case ever happen? Is _ToOneState.unstored not only set in the `target` setter on the pre-condition that the ID is 0?

Reply:
happens here: https://github.com/objectbox/objectbox-dart/pull/162/files#diff-7f77c39ca497ed9b247c62f1f433e1d33f2ec7e03f0450b657b9d367584c0a0fR362 
```dart
void _putToOneRelFields(T object, PutMode mode) {
    _entity.toOneRelations(object).forEach((ToOne rel) {
      // put new objects
      if (rel.hasValue && rel.targetId == 0) {
        rel.targetId = rel.internalTargetBox._put(rel.target, mode, true);
      }
    });
  }
  ```
  
Or rather it will happen once put() actually calls `setId()` on new objects (that's already implemented in the optimizations PR #164 - https://github.com/objectbox/objectbox-dart/pull/164/files#diff-7f77c39ca497ed9b247c62f1f433e1d33f2ec7e03f0450b657b9d367584c0a0fR101)

Reply:
Ah, thanks, that makes sense. Maybe add a comment to refer to that put case (or put differently add the reason why this condition should exist)?

Reply:
added a comment:
```
      // Optimization for targetId being set from box.put(sourceObject)
      // after entity.setId(object, newID) was already called on the new target.
```